Implement options page to persist Prettier config

nickmccurdy opened the pull request:
Fixes #25

I still need to improve the UI, but feedback is welcome

![image](https://user-images.githubusercontent.com/927220/67299247-65c4c700-f4ba-11e9-8af2-bb800ec8956e.png)

kaicataldo commented on extension/options/src/App.js:
```diff
@@ -0,0 +1,140 @@
+  const [options, setOptions] = useState();
+
+  useEffect(() => {
+    window.chrome.storage.sync.get(defaultOptions, setOptions);
```
Do we want to scope these settings to a property (`config`, maybe?) in case we want to add other types of data in here later? That way we wouldn't have to make this code backwards compatible in the future.

nickmccurdy replied:
That's a good point, I wasn't thinking about this. It seems like we could store all the options under one object and manually merge them in.

kaicataldo commented on extension/options/src/App.js:
```diff
@@ -0,0 +1,140 @@
+  }, [setOptions]);
+
+  useEffect(() => {
+    if (options) {
```
When would `options` evaluate to falsy here?

nickmccurdy replied:
`setOptions` is asyncronous, so the `options` state isn't updated until the next render, and remains `undefined` on the first call of this `useEffect` hook.

kaicataldo replied:
Ah, got it! Thanks for the explanation!

kaicataldo commented on extension/options/src/App.js:
```diff
@@ -0,0 +1,140 @@
+    });
+  }
+
+  return options ? (
```
When would `options` evaluate to falsy here?

nickmccurdy replied:
https://github.com/prettier/prettier-chrome-extension/pull/56#discussion_r337854482

kaicataldo commented on .eslintrc.json:
```diff
@@ -2,7 +2,7 @@
     "node": true
   },
-  "plugins": ["prettier"],
+  "plugins": ["prettier", "react"],
```
Should we turn on the recommended rule set?

kaicataldo commented on extension/options/src/App.js:
```diff
@@ -0,0 +1,140 @@
+  return options ? (
+    <>
+      <h1>
+        <a href="https://prettier.io/docs/en/options.html">Prettier Options</a>
```
Might be good to turn the [`jsx-no-target-blank`]( https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-no-target-blank.md) rule on.

nickmccurdy replied:
Is there a rule that bans `target="_blank"` completely (not just when it's insecure)? Screenreaders don't consistently implement support for this, so the extra tab/page can cause accessibility issues.

nickmccurdy replied:
Is there a rule that bans `target="_blank"` completely (not just when it's insecure)? Screenreaders don't consistently implement support for this, so the extra tab/page can cause accessibility issues.

kaicataldo replied:
Whoops, I thought this had the `target="_blank"` attribute. In this case, might it make sense to add it since this will be any options pop up in Chrome?

nickmccurdy replied:
I'm not sure how links are handled in popups, so I'm trying to test it out. I have this in the manifest (instead of `options_page`) and it's not letting me open the popup, any ideas? ```js
  "options_ui": {
    "page": "options/build/index.html",
    "open_in_tab": false
  },
```

kaicataldo replied:
Hmmm I'm not sure - I haven't actually made a Chrome Extension settings page before. The popup isn't showing up at all?

nickmccurdy replied:
Nope, the button is clicked and nothing shows up, not even an error

nickmccurdy replied:
I still can't get this to work. I'm giving up for now but willing to use a popup if someone else figures it out.

kaicataldo commented on extension/options/src/App.js:
```diff
@@ -0,0 +1,140 @@
+  const [options, setOptions] = useState();
+
+  useEffect(() => {
+    window.chrome.storage.sync.get(defaultOptions, setOptions);
```
Seems like we only want to `chrome.storage.sync.get` this data on initial mount, and then we want to `chrome.storage.sync.set` every time time `options` changes (we can rely on this component's internal state for the view after the initial load). What does watching for `setOptions` do here? I'm new to hooks, so might be misunderstanding something :)

nickmccurdy replied:
`eslint-config-react-app` (part of CRA) complains if values are used in `useEffect` without being added to the dependencies array. I'm not sure if it really matters in this case: because the `setOptions` reference is stable, it will only run on the initial mount anyway (as if the dependencies were `[]` or you were using `componentDidMount` in a class component).

kaicataldo replied:
Got it, thanks! I think it’s clearer to have it be an empty array, but I’m also new to hooks and defer to your judgment here.

nickmccurdy replied:
Whoops I must have been thinking about some previous code, it does actually allow the dependencies here to be `[]` because it knows `setOptiosn` is stable, updated.

nickmccurdy commented:
> ... I don't think this needs to be a completely separate app. Can we just update the webpack config to output a second bundle? We don't need to compile anything except JSX, I think, so adding and configuring Babel should be fairly straightforward.

I was originally hoping that CRA would make the build setup easier, but it's kind of a hassle to manually run two different build scripts depending on which files change. I'll give it a shot.

nickmccurdy commented:
Travis seems to be stalling again, I don't even see these commits in the PR builds 😢 
![image](https://user-images.githubusercontent.com/927220/67371020-afa8be00-f549-11e9-9efb-c727d4444480.png)

lipis commented:
Seems to be OK now.. or?

kaicataldo commented on extension/options/public/index.html:
```diff
@@ -0,0 +1,12 @@
+<!DOCTYPE html>
```
I think it makes sense to just move this to the root of the `extension` directory (unless we want to create a `public` directory there?).

nickmccurdy replied:
It's now in `extension/src/options/index.html`. Since it's only used for options, I don't think it feels right to have it at the top level.

kaicataldo replied:
If it's going to live in `src`, can we use [CopyWebpackPlugin](https://webpack.js.org/plugins/copy-webpack-plugin) to copy it into `dist`?

nickmccurdy replied:
That could work, but I feel like it's only worth it if we're just publishing `dist`.

kaicataldo replied:
To be clearer, I don't think we should be shipping anything in `src`. If things don't need to be run through webpack, I would advocate for either using the webpack plugin above or putting them in a directory like `extension/{public,assets}`.

nickmccurdy replied:
I see what you mean. Would you prefer to do this before merging/publishing the options page?

kaicataldo replied:
I’m happy to implement it after.

kaicataldo commented on extension/options/src/App.js:
```diff
@@ -0,0 +1,140 @@
+import React, { useEffect, useState } from "react";
```
Can these files go in `extension/src/options/*`?

nickmccurdy replied:
I read your mind, just did this before reading your comment 😎

kaicataldo commented on webpack.config.js:
```diff
@@ -2,8 +2,29 @@ const path = require("path");
+        exclude: /node_modules/,
+        test: /\.js$/,
+        use: {
+          loader: "babel-loader",
```
🙌

nickmccurdy commented:
I implemented a custom webpack config and moved the files around a bit. Do we like this setup? Alternatively, how would we feel about moving the contents of `extension` to the root and publishing `dist` instead?

kaicataldo commented:
I'm not against moving it into the root, but I do think we want to make sure we're only publishing the files needed for the extension, which means we'll need to pick and choose which files to include (`manifest.json/`, `dist/`, `icons/`)

nickmccurdy commented:
That's what I was thinking of, but I'm not sure of the easiest way to implement it with webpack. I feel like what we have is okay until we start putting larger files in `extension/src` though.

kaicataldo commented:
I'm happy to do that work with webpack once we land this :)

kaicataldo commented on .eslintrc.json:
```diff
@@ -52,7 +49,9 @@
-    "strict": ["error", "global"]
+    "strict": ["error", "global"],
+    "react-hooks/rules-of-hooks": "error",
+    "react-hooks/exhaustive-deps": "warn"
```
Can we set this to `"error"` as well? Seems like we would want to block merging for these rules.

nickmccurdy commented:
This will turn into a normal mergeable PR once we click "Ready for review", but I still have some things I'd like to fix first.

Do you have any preferences on making the UI pretty before or after merging this? I'd be fine either way.

kaicataldo commented:
Whatever you prefer!

lipis commented:
let's make it work.. then make it better :)

kaicataldo commented:
I think this is a good foundation for the settings page, and once we change [this](https://github.com/prettier/prettier-chrome-extension/pull/56#discussion_r337850394) we could merge this and follow up with styling and moving files around!

nickmccurdy commented:
Yea sorry, still working on that. I ran into some issues and set up devtools to make it easier to troubleshoot, but still need to finish tweaking its configuration.

nickmccurdy commented:
Now that #59 is merged (which was an interactive rebase of this branch), should I rebase this on top of master?

lipis commented:
No need  to rebase.. we are squashing the PRs anyways.

nickmccurdy commented:
I figured out how to scope the `options` property in Chrome's extension storage. I can help out with publishing and UI fixes after this is merged.

lipis commented:
I couldn't find a way to add more admins to the extension dashboard.. or how could I transfer it to another account so we can enable the automatic publishing.

nickmccurdy commented:
To be clear, I don't need publishing access for this PR specifically, as the storage is automatically set up with the user's Chrome storage and Google account (if any) due to the `storage` permission in `package.json`.

If you want to set up group publishing see [Set up Group Publishing](https://developer.chrome.com/webstore/publish#set-up-group-publishing). Unfortunately this process is irreversible and charges each new publisher $5, but we could try auto publishing with a single account if you prefer (which I think I would too).

kaicataldo commented on extension/src/content/index.js:
```diff
@@ -482,4 +486,4 @@ function init() {
 }
 
-init();
+chrome.storage.sync.get(init);
```
This isn't necessarily a blocker (we can fix this later if need be), but this only runs when the page is loaded, correct? One use case this might not handle very gracefully is when a GitHub/StackOverflow tab is open and someone goes and changes the settings (they wouldn't go into effect until the page is reloaded).

We could probably fix this by reading from chrome storage each time Prettier formats (again, does not need to be a blocker to land this, just want to make sure it's commented here for posterity).

nickmccurdy replied:
I forgot to implement this, but we should be able to use [`chrome.storage.onChanged.addListener `](https://developer.chrome.com/extensions/storage#event-onChanged)

lipis commented:
Published in 0.0.4
